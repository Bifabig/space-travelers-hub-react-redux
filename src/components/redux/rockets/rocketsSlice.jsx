import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const getRocketsURL = 'https://api.spacexdata.com/v3/rockets';

const initialState = {
  rockets: [],
  isLoading: false,
  error: undefined,
};

export const getRockets = createAsyncThunk('rockets/getRockets', async (thunkAPI) => {
  try {
    const response = await axios(getRocketsURL);
    return response.data;
  } catch (err) {
    return thunkAPI.rejectWithValue(err);
  }
});

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    reserveRocket: (state, action) => {
      const rocketId = action.payload;
      const newState = state;
      newState.rockets = state.rockets.map((rocket) => (
        rocket.rocket_id !== rocketId ? rocket : { ...rocket, reserved: true }
      ));
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getRockets.pending, (state) => {
        const pendingState = state;
        pendingState.isLoading = true;
      })
      .addCase(getRockets.fulfilled, (state, action) => {
        const fulfilledState = state;
        fulfilledState.isLoading = false;
        fulfilledState.rockets = action.payload;
        console.log(fulfilledState.rockets);
      })
      .addCase(getRockets.rejected, (state) => {
        const rejectedState = state;
        rejectedState.isLoading = false;
      });
  },
});

export const { reserveRocket, remove } = rocketsSlice.actions;
export default rocketsSlice.reducer;
