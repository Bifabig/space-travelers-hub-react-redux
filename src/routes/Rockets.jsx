const Books = () => (
  <div className="book-form-wrapper">
    <div className="books-wrapper">
      Hello
    </div>
  </div>
);
export default Books;
