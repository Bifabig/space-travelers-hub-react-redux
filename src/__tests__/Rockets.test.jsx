import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import Rockets from '../routes/Rockets';
import { getRockets, reserveRocket, cancelReserveRocket } from '../components/redux/rockets/rocketsSlice';
import '@testing-library/jest-dom/extend-expect';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../components/redux/rockets/rocketsSlice', () => ({
  getRockets: jest.fn(),
  reserveRocket: jest.fn(),
  cancelReserveRocket: jest.fn(),
}));

describe('Rockets', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useSelector.mockReturnValue({
      rockets: [],
      isLoading: false,
      error: null,
    });
  });

  it('should render loading message when isLoading is true', () => {
    useSelector.mockReturnValue({
      rockets: [],
      isLoading: true,
      error: null,
    });

    render(<Rockets />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('should render error message when error exists', () => {
    useSelector.mockReturnValue({
      rockets: [],
      isLoading: false,
      error: 'Something went wrong',
    });

    render(<Rockets />);

    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    const { container } = render(<Rockets />);
    expect(container).toMatchSnapshot();
  });

  it('should dispatch getRockets when rockets array is empty', () => {
    useSelector.mockReturnValue({
      rockets: [],
      isLoading: false,
      error: null,
    });

    const { container } = render(<Rockets />);

    expect(getRockets).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  it('should render the rockets', () => {
    useSelector.mockReturnValue({
      rockets: [
        {
          rocket_id: '1',
          rocket_name: 'Falcon 9',
          flickr_images: ['image_url'],
          reserved: false,
          description: 'Rocket description',
        },
      ],
      isLoading: false,
      error: null,
    });

    const { container } = render(<Rockets />);

    expect(screen.getByText('Falcon 9')).toBeInTheDocument();
    expect(screen.getByText('Rocket description')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should dispatch reserveRocket when clicking on "Reserve Rocket" button', () => {
    const rocketId = '1';
    useSelector.mockReturnValue({
      rockets: [
        {
          rocket_id: rocketId,
          rocket_name: 'Falcon 9',
          flickr_images: ['image_url'],
          reserved: false,
          description: 'Rocket description',
        },
      ],
      isLoading: false,
      error: null,
    });

    render(<Rockets />);

    const reserveButton = screen.getByText('Reserve Rocket');
    fireEvent.click(reserveButton);

    expect(reserveRocket).toHaveBeenCalledWith(rocketId);
  });

  it('should dispatch cancelReserveRocket when clicking on "Cancel Reservation" button', () => {
    const rocketId = '1';
    useSelector.mockReturnValue({
      rockets: [
        {
          rocket_id: rocketId,
          rocket_name: 'Falcon 9',
          flickr_images: ['image_url'],
          reserved: true,
          description: 'Rocket description',
        },
      ],
      isLoading: false,
      error: null,
    });

    render(<Rockets />);

    const cancelButton = screen.getByText('Cancel Reservation');
    fireEvent.click(cancelButton);

    expect(cancelReserveRocket).toHaveBeenCalledWith(rocketId);
  });
});
