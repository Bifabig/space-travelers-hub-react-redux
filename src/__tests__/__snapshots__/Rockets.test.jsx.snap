// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets should dispatch getRockets when rockets array is empty 1`] = `
<div>
  <hr />
  <div
    class="rocketsWrapper"
  />
</div>
`;

exports[`Rockets should render error message when error exists 1`] = `
<div>
  <h2>
    Something went wrong
  </h2>
</div>
`;

exports[`Rockets should render the rockets 1`] = `
<div>
  <hr />
  <div
    class="rocketsWrapper"
  >
    <div
      class="rocket"
    >
      <img
        alt=""
        class="image"
        src="image_url"
      />
      <div
        class="rocketInfo"
      >
        <h2>
          Falcon 9
        </h2>
        <p
          class="rocketDescription"
        >
          Rocket description
        </p>
        <button
          class="btn btn-primary"
          type="button"
        >
          Reserve Rocket
        </button>
      </div>
    </div>
  </div>
</div>
`;
